src/vmm/src/devices/virtio/block/virtio/request.rs

```diff
@@ -313,6 +334,52 @@ impl Request {
                     return Err(VirtioBlockError::InvalidDataLength);
                 }
             }
+            RequestType::Discard => {
+                // Get data descriptor
+                let data_desc = avail_desc.next_descriptor()
+                    .ok_or(VirtioBlockError::DescriptorChainTooShort)?;
+                if data_desc.is_write_only() {
+                    return Err(VirtioBlockError::UnexpectedWriteOnlyDescriptor);
+                }
+            
+                // Validate data length
+                let segment_size = std::mem::size_of::<DiscardSegment>() as u32; // 16 bytes
+                if data_desc.len % segment_size != 0 {
+                    return Err(VirtioBlockError::InvalidDataLength);
+                }
+            
+                // Calculate number of segments
+                let num_segments = data_desc.len / segment_size;
+                if num_segments == 0 {
+                    return Err(VirtioBlockError::InvalidDiscardRequest);
+                }
+                let mut segments = Vec::with_capacity(num_segments as usize);
```

Comment:
I think we can avoid having to deal with multiple segments if we just limit their number to 1 in the config space of the device (`max_discard_sectors`). But it will need to be defined first as currently we treat it just as a `Vec<u8>`. I can help you with this. Also in all other code please avoid dynamic allocations, it can cause issues in the future.